src/shims/unix/mem.rs

```diff
@@ -0,0 +1,244 @@
+        if length == 0 {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
+            return Ok(Scalar::from_maybe_pointer(Pointer::null(), this));
+        }
+
+        // If a user tries to map a file, we want to loudly inform them that this is not going
+        // to work. It is possible that POSIX gives us enough leeway to return an error, but the
+        // outcome for the user (I need to add cfg(miri)) is the same, just more frustrating.
+        if fd != -1 {
+            throw_unsup_format!("Miri does not support file-backed memory mappings");
+        }
+
+        // POSIX says:
+        // [ENOTSUP]
+        // * MAP_FIXED or MAP_PRIVATE was specified in the flags argument and the implementation
+        // does not support this functionality.
+        // * The implementation does not support the combination of accesses requested in the
+        // prot argument.
+        //
+        // Miri doesn't support MAP_FIXED or any any protections other than PROT_READ|PROT_WRITE.
+        if flags & map_fixed != 0 || prot != prot_read | prot_write {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("ENOTSUP")))?;
+            return Ok(Scalar::from_maybe_pointer(Pointer::null(), this));
+        }
+
+        // Miri does not support shared mappings, or any of the other extensions that for example
+        // Linux has added to the flags arguments.
+        if flags != map_private | map_anonymous {
+            throw_unsup_format!(
+                "Miri only supports calls to mmap which set the flags argument to MAP_PRIVATE|MAP_ANONYMOUS"
+            );
+        }
+
+        // This is only used for file mappings, which we don't support anyway.
+        if offset != 0 {
+            throw_unsup_format!("Miri does not support non-zero offsets to mmap");
+        }
+
+        let align = Align::from_bytes(this.machine.page_size).unwrap();
+        let map_length = this.machine.round_up_to_multiple_of_page_size(length).unwrap_or(u64::MAX);
+
+        let ptr =
+            this.allocate_ptr(Size::from_bytes(map_length), align, MiriMemoryKind::Mmap.into())?;
+        // We just allocated this, the access is definitely in-bounds and fits into our address space.
+        // mmap guarantees new mappings are zero-init.
+        this.write_bytes_ptr(
+            ptr.into(),
+            std::iter::repeat(0u8).take(usize::try_from(map_length).unwrap()),
+        )
+        .unwrap();
+        // Memory mappings are always exposed
+        Machine::expose_ptr(this, ptr)?;
+
+        Ok(Scalar::from_pointer(ptr, this))
+    }
+
+    fn mremap(
+        &mut self,
+        old_address: &OpTy<'tcx, Provenance>,
+        old_size: &OpTy<'tcx, Provenance>,
+        new_size: &OpTy<'tcx, Provenance>,
+        flags: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Scalar<Provenance>> {
+        let this = self.eval_context_mut();
+
+        let old_address = this.read_pointer(old_address)?;
```

Comment:
You said in comments that provenance shouldn't matter here. In that case you should do `read_target_usize` and then `ptr_from_addr_cast`.